rfcs/023-images-endpoint/README.md

```diff
@@ -0,0 +1,48 @@
+# RFC 023: Images endpoint
+
+**Status:** :construction: Draft :building_construction:
+
+**Last updated:** 27/04/2020
+
+## Context and requirements
+
+- Images are now entities in the catalogue API
+- We want to be able to search for images, and to fetch specified images
+- The images endpoint should behave similarly to the works endpoint
+
+## Proposal
+
+The initial images endpoint will be in **beta**, and therefore will be liable to have breaking changes made without a version bump.
+
+### Image display model
+
+```
+{
+  "id": String,
+  "location": DigitalLocation,
```

Comment:
I think this should potentially be `locations` and a list. That way we can support eg a IIIF image location and a location that is a deep link into the IIIF presentation API.

Reply:
The locations should also copy across license, credit, access conditions etc I think. You should be able to do the right thing here without having to refer to the works endpoint.

Reply:
iirc we discussed this before and decided on 1 location per image? I'm wary of this as it's not just an API change but also a model change with huge consequences (eg, you can't give an `Image` that contains multiple images a feature vector).

I think that last time we said that stuff like deeplinks could be in a separate field to avoid this problem (which also wouldn't be a breaking API change to add).

Reply:
> The locations should also copy across license, credit, access conditions etc I think. You should be able to do the right thing here without having to refer to the works endpoint.

Yes, this already happens, `DigitalLocation` is a shorthand for the exact same schema as in the works API.

Reply:
I don't mean that an image is multiple images, I mean that an image can potentially exist in multiple locations. Difference between an image as something conceptual and some concrete (eg an actual jpg).

Reply:
Basically I think multiple locations is good, having that expressed in a flat list is bad.

Reply:
> I don't mean that an image is multiple images, I mean that an image can potentially exist in multiple locations.

Right, but you still need to have one of these as canonical for inference etc

Reply:
You are correct that better describing is better, we haven't done that on `/works` though. Though there are more kinds of locations there.

Reply:
You should be able to tell from the locationType what kind of location it is, though you are correct that doesn't describe the relationship.

Reply:
I'm saying that there is a concrete problem with having a `locations` list: which one of those does the inferrer fetch to extract feature vectors from?

Reply:
The iiif image, if you want the actual physical image

Reply:
Does that not mean that `locations` a list of 2+ different types of things, with an implicit constraint that there is only 1 thing of 1 of the types? In which case we should make that explicit eg:
```
  "location": blah,
  "presentation": blah
```

Reply:
I'm not sure I want to tie the model to location types so strongly, but would definitely like input from others here esp with respect to the same situation on works.

Reply:
> I'm saying that there is a concrete problem with having a `locations` list: which one of those does the inferrer fetch to extract feature vectors from?

Can we not have differing internal & internal models where the canonical location for the inferrer is explicit internally but hidden from the API consumer?

Reply:
> You are correct that better describing is better, we haven't done that on /works though. Though there are more kinds of locations there.

Personally I would have really preferred more explicit relationships on `works`, as a client and someone writing code in the pipeline.

---

I like @kenoir's suggestion.

A similar example would be thumbnails on `Work`s.

It is just a digital location, but special as we have given it a context or way for it to be used.

Could this just be a case of having `locations` which works the same as in `works`.

We would have the `vectorsFrom: DigitalLocation` as internal as I can't imagine a client finding this useful.

Having a `thumbnail` type param, or better suggested `presentation` or `display` (preferably with a URL template that can be used for the client to create it's own sizes) would also be super useful.

I can't see a problem with there being duplication in the more explicitly related locations and the list.

If we just have the more explicit locations, adding the the more generic `locations` list later should be easy?

Reply:
Ok let's look at being explicit then, we actually need to change locations on works as part of the locations and requesting work anyway, so this might lead us towards a pattern that we can reuse there.